src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,137 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int INITIAL_CAPACITY = 10;
+    private T[] values;
+    private int elementsAdded;
+    private boolean resize;
+
+    public ArrayList() {
+        values = (T[])new Object[INITIAL_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        int index = checkIndex();
+        if (index >= values.length) {
+            resize = true;
+            resize();
+        }
+        values[index] = value;
+        elementsAdded++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index > checkIndex() || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index " + index
+                    + " out of bounds for length " + size());
+        }
+        if (index == checkIndex()) {
+            if (index >= values.length) {
+                resize = true;
+                resize();
+            }
+            add(value);
+        } else {
+            if (size() == values.length) {
+                resize = true;
+                resize();
+            }
+            T[] tmp = (T[]) new Object[size()];
+            System.arraycopy(values, index, tmp, 0, size() - index);
+            values[index] = value;
+            System.arraycopy(tmp, 0, values, index + 1, size() - index);
+            elementsAdded++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index >= 0 && index < size()) {
+            return values[index];
+        }
+        throw new ArrayListIndexOutOfBoundsException("Index " + index
+                + " out of bounds for length " + size());
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (size() == 0 || index >= checkIndex() || index < 0) {
```

Comment:
```suggestion
        if (index < 0 || index >= size) {
```

and lets move logic of checking index in separate private method (as we use it several times in code)